apps/src/lab2/levelEditors/aiCustomizations/EditAiCustomizations.tsx

```diff
@@ -37,75 +81,161 @@ const EditAiCustomizations: React.FunctionComponent<
+        },
+      });
+    },
+    [aiCustomizations, setAiCustomizations]
+  );
+
   return (
-    <div>
-      <input
-        type="hidden"
-        id="level_initial_ai_customizations"
-        name="level[initial_ai_customizations]"
-        value={JSON.stringify(aiCustomizations)}
-      />
-      <BodyThreeText>
-        Set the initial values and visibility for AI model customizations.
-        <br />
-        <br />
-        <b>Editable:</b> students can change the value.
-        <br />
-        <b>Read Only:</b> students can see the value but cannot change it.
-        <br />
-        <b>Hidden:</b> the field is not shown on the customization panel.
-      </BodyThreeText>
-      <div className={moduleStyles.fieldSection}>
-        <BodyOneText>Bot Name</BodyOneText>
-        <div className={moduleStyles.fieldRow}>
-          <div className={moduleStyles.fieldValue}>
-            <label htmlFor="botName">Initial Value</label>
-            <input
-              id="botName"
-              type="text"
-              value={aiCustomizations.botName?.value || ''}
-              onChange={e =>
+    <UpdateContext.Provider
+      value={{
+        aiCustomizations,
+        setPropertyVisibility,
+        setPropertyValue,
+        setModelCardPropertyValue,
+      }}
+    >
+      <div>
+        <input
+          type="hidden"
+          id="level_initial_ai_customizations"
+          name="level[initial_ai_customizations]"
+          value={JSON.stringify(sanitizeData(aiCustomizations))}
+        />
+        <BodyThreeText>
+          Set the initial values and visibility for AI model customizations.
+          <br />
+          <br />
+          <b>Editable:</b> students can change the value.
+          <br />
+          <b>Read Only:</b> students can see the value but cannot change it.
+          <br />
+          <b>Hidden:</b> the field is not shown on the customization panel.
+        </BodyThreeText>
+        <FieldSection
+          fieldName="botName"
+          labelText="Bot Name"
+          inputType="text"
+          description="Name of the chat bot."
+        />
+        <FieldSection
+          fieldName="temperature"
+          labelText="Temperature"
+          inputType="number"
+          min={MIN_TEMPERATURE}
+          max={MAX_TEMPERATURE}
+          step={0.1}
+          description="Temperature setting for the model. A higher temperature induces more randomness."
+        />
+        <FieldSection
+          fieldName="systemPrompt"
+          labelText="System Prompt"
+          description="The base prompt applied to all inputs to the model."
+          inputType="textarea"
+        />
+        <FieldSection
+          fieldName="retrievalContexts"
+          labelText="Retrieval Contexts"
+          description="Pieces of information the model references when generating a response."
+          inputType="custom"
+          inputNode={
+            <MultiItemInput
```

Comment:
Nit: if there's a way to reduce the default vertical space for these textboxes for retrieval specifically, I think I'd prefer that. I'd rather be able to see most of the retrieval items together on the screen rather than scroll to view them, but the current vertical default makes me think I'll quickly get overloaded with vertical space if I add too many.

I guess to test that, what would it look like with the 5 retrieval sentences I added in my Tucson bot prototype? How much vertical space is added, and how much of it is mostly empty?
![image](https://github.com/code-dot-org/code-dot-org/assets/68714964/4d2f7e6c-de18-416b-838f-69743e556d26)

Reply:
Tweaked this a bit to try to take more advantage of horizontal space and reduce vertical space. Thoughts? Also worth noting that all of these text boxes are resizable. I'll also follow up with a separate change to make these sections collapsible.
<img width="1512" alt="Screenshot 2024-03-08 at 11 47 31 AM (2)" src="https://github.com/code-dot-org/code-dot-org/assets/85528507/2eea66c8-abf6-4db5-bc06-b58cdfa085e4">